src/serializers/type_serializers/union.rs

```diff
@@ -198,24 +202,180 @@ impl TypeSerializer for UnionSerializer {
+                }
             }
         }
-        UnionSerializer::from_choices(choices)
+
+        let basic_union_ser = UnionSerializer::from_choices(self.choices.clone());
+        if let Ok(s) = basic_union_ser {
+            return s.json_key(key, extra);
+        }
+
+        extra.warnings.on_fallback_py(self.get_name(), key, extra)?;
+        infer_json_key(key, extra)
+    }
+
+    fn serde_serialize<S: serde::ser::Serializer>(
+        &self,
+        value: &Bound<'_, PyAny>,
+        serializer: S,
+        include: Option<&Bound<'_, PyAny>>,
+        exclude: Option<&Bound<'_, PyAny>>,
+        extra: &Extra,
+    ) -> Result<S::Ok, S::Error> {
+        let py = value.py();
+        let mut new_extra = extra.clone();
+        new_extra.check = SerCheck::Strict;
+
+        if let Some(tag) = self.get_discriminator_value(value) {
+            let tag_str = tag.to_string();
+            if let Some(selected_serializer) = self.lookup.get(&tag_str) {
+                match selected_serializer.to_python(value, include, exclude, &new_extra) {
+                    Ok(v) => return infer_serialize(v.bind(py), serializer, None, None, extra),
+                    Err(_) => {
+                        if self.retry_with_lax_check() {
+                            new_extra.check = SerCheck::Lax;
+                            match selected_serializer.to_python(value, include, exclude, &new_extra) {
+                                Ok(v) => return infer_serialize(v.bind(py), serializer, None, None, extra),
+                                Err(err) => return Err(py_err_se_err(err)),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        let basic_union_ser = UnionSerializer::from_choices(self.choices.clone());
+        if let Ok(s) = basic_union_ser {
+            return s.serde_serialize(value, serializer, include, exclude, extra);
+        }
+
+        extra.warnings.on_fallback_ser::<S>(self.get_name(), value, extra)?;
+        infer_serialize(value, serializer, include, exclude, extra)
+    }
+
+    fn get_name(&self) -> &str {
+        &self.name
+    }
+}
+
+impl TaggedUnionSerializer {
+    fn get_discriminator_value(&self, value: &Bound<'_, PyAny>) -> Option<Py<PyAny>> {
+        let py = value.py();
+        match &self.discriminator {
+            Discriminator::LookupKey(lookup_key) => match lookup_key {
+                LookupKey::Simple { py_key, .. } => value.getattr(py_key).ok().map(|obj| obj.to_object(py)),
+                _ => None,
+            },
+            Discriminator::Function(func) => func.call1(py, (value,)).ok().or_else(|| {
+                // Try converting object to a dict, might be more compatible with poorly defined callable discriminator
+                value
+                    .call_method0(intern!(py, "dict"))
+                    .and_then(|v| func.call1(py, (v.to_object(py),)))
+                    .ok()
```

Comment:
I don't love this much, where does the `.dict()` method come from? I guess we're expecting this to be a `BaseModel`? What about dataclasses etc? Maybe better to not have this fallback and use it as a nudge to encourage users to fix their discriminators...

Reply:
Yeah, this makes things... slow. The idea is that most people design callable discriminators around the idea that they're accepting dictionary type input data (not great, but lots of folks do it), so I think it's worth trying this fallback.

Reply:
Especially if it's a performance problem, it doesn't really feel like we're helping users by having this fallback then (it just allows them to accidentally write bad code).

One option is that we could maybe (as a one-off) try this and see if it succeeds, if it does, then we can emit a warning saying the discriminator function is poorly implemented.

I think if the tag function is failing we definitely want to emit that exception somehow, so users can debug.